warden/x/act/keeper/actions_keeper.go

```diff
@@ -99,3 +101,35 @@ func (k ActionKeeper) Import(ctx sdk.Context, actions []types.Action) error {
 func (k ActionKeeper) Coll() repo.SeqCollection[types.Action] {
 	return k.actions
 }
+
+func (k ActionKeeper) PruneAction(ctx context.Context, action types.Action) error {
+	if err := k.actions.Remove(ctx, action.Id); err != nil {
+		return err
+	}
+
+	for _, addr := range action.Mentions {
+		key := collections.Join(sdk.MustAccAddressFromBech32(addr), action.Id)
+		if err := k.actionByAddress.Remove(ctx, key); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (k ActionKeeper) ExpiredActions(ctx context.Context, statuses map[types.ActionStatus]struct{}, blockTime time.Time, timeout time.Duration) ([]types.Action, error) {
+	isExceededTimeout := func(key uint64, value types.Action) (include bool, err error) {
+		_, isCompletedStatus := statuses[value.Status]
+		isKeepTimeExceeded := blockTime.After(value.UpdatedAt.Add(timeout))
+		return isCompletedStatus && isKeepTimeExceeded, nil
+	}
+
+	id := func(key uint64, value types.Action) (types.Action, error) {
+		return value, nil
+	}
+
+	actions, _, err := query.CollectionFilteredPaginate[uint64, types.Action, repo.SeqCollection[types.Action], types.Action](
+		ctx, k.Coll(), nil, isExceededTimeout, id)
+
+	return actions, err
+}
```

Comment:
i think that we can simplify this function to embed the logic of "what are the expired actions" instead of taking the statuses a param

this is a non-optimized version that will blindly iterate over all actions and load all the expired actions in memory - i believe that this is okay for now as the number of actions will stabilize after the first pruning, but for the future we'll need a queue based on the UpdatedAt field (the staking modules do something like that)

```suggestion
func (k ActionKeeper) ExpiredActions(ctx context.Context, pendingTimeout, completedTimeout time.Duration) ([]types.Action, error) {
	it, err := k.Coll().Iterate(ctx, nil)
	if err != nil {
		return nil, err
	}
	defer it.Close()

	var actions []types.Action
	for ; it.Valid(); it.Next() {
		act, err := it.Value()
		if err != nil {
			return nil, err
		}

		if isExpired(ctx, act, pendingTimeout, completedTimeout) {
			actions = append(actions, act)
		}
	}

	return actions, nil
}

func isExpired(ctx context.Context, act types.Action, pendingTimeout, completedTimeout time.Duration) bool {
	if act.Status == types.ActionStatus_ACTION_STATUS_TIMEOUT {
		return true
	}

	now := sdk.UnwrapSDKContext(ctx).HeaderInfo().Time

	if act.Status == types.ActionStatus_ACTION_STATUS_PENDING && act.UpdatedAt.Add(pendingTimeout).Before(now) {
		return true
	}

	if act.Status == types.ActionStatus_ACTION_STATUS_COMPLETED && act.UpdatedAt.Add(completedTimeout).Before(now) {
		return true
	}

	return false
}
```